libglusterfs/src/run_fork.c

```diff
@@ -0,0 +1,576 @@
+}
+
+void
+runner_add_args(runner_t *runner, ...)
+{
+    va_list argp;
+
+    va_start(argp, runner);
+    runner_va_add_args(runner, argp);
+    va_end(argp);
+}
+
+void
+runner_argprintf(runner_t *runner, const char *format, ...)
+{
+    va_list argva;
+    char *arg = NULL;
+    int ret = 0;
+
+    va_start(argva, format);
+    ret = gf_vasprintf(&arg, format, argva);
+    va_end(argva);
+
+    if (ret < 0) {
+        runner->runerr = errno;
+        return;
+    }
+
+    runner_insert_arg(runner, arg);
+}
+
+void
+runner_log(runner_t *runner, const char *dom, gf_loglevel_t lvl,
+           const char *msg)
+{
+    char *buf = NULL;
+    size_t len = 0;
+    int i = 0;
+
+    if (runner->runerr)
+        return;
+
+    for (i = 0;; i++) {
+        if (runner->argv[i] == NULL)
+            break;
+        len += (strlen(runner->argv[i]) + 1);
+    }
+
+    buf = GF_CALLOC(1, len + 1, gf_common_mt_run_logbuf);
+    if (!buf) {
+        runner->runerr = errno;
+        return;
+    }
+    for (i = 0;; i++) {
+        if (runner->argv[i] == NULL)
+            break;
+        strcat(buf, runner->argv[i]);
+        strcat(buf, " ");
+    }
+    if (len > 0)
+        buf[len - 1] = '\0';
+
+    gf_msg_callingfn(dom, lvl, 0, LG_MSG_RUNNER_LOG, "%s: %s", msg, buf);
+
+    GF_FREE(buf);
+}
+
+void
+runner_redir(runner_t *runner, int fd, int tgt_fd)
+{
+    GF_ASSERT(fd > 0 && fd < 3);
+
+    if ((fd > 0) && (fd < 3))
+        runner->chfd[fd] = (tgt_fd >= 0) ? tgt_fd : -2;
+}
+
+int
+runner_start(runner_t *runner)
+{
+    int pi[3][2] = {{-1, -1}, {-1, -1}, {-1, -1}};
+    int xpi[2];
+    int ret = 0;
+    int errno_priv = 0;
+    int i = 0;
+    sigset_t set;
+
+    if (runner->runerr || !runner->argv) {
+        errno = (runner->runerr) ? runner->runerr : EINVAL;
+        return -1;
+    }
+
+    GF_ASSERT(runner->argv[0]);
+
+    /* set up a channel to child to communicate back
+     * possible execve(2) failures
+     */
+    ret = pipe(xpi);
+    if (ret != -1)
+        ret = fcntl(xpi[1], F_SETFD, FD_CLOEXEC);
+
+    for (i = 0; i < 3; i++) {
+        if (runner->chfd[i] != -2)
+            continue;
+        ret = pipe(pi[i]);
+        if (ret != -1) {
+            runner->chio[i] = fdopen(pi[i][i ? 0 : 1], i ? "r" : "w");
+            if (!runner->chio[i])
+                ret = -1;
+        }
+    }
+
+    if (ret != -1)
+        runner->chpid = fork();
+    switch (runner->chpid) {
+        case -1:
+            errno_priv = errno;
+            sys_close(xpi[0]);
+            sys_close(xpi[1]);
+            for (i = 0; i < 3; i++) {
+                sys_close(pi[i][0]);
+                sys_close(pi[i][1]);
+            }
+            errno = errno_priv;
+            return -1;
+        case 0:
+            for (i = 0; i < 3; i++)
+                sys_close(pi[i][i ? 0 : 1]);
+            sys_close(xpi[0]);
+            ret = 0;
+
+            for (i = 0; i < 3; i++) {
+                if (ret == -1)
+                    break;
```

Comment:
Is it not the control flow construct that your kind self spotted in #1902 to irk Coverity? :P

Please adjust the resurrected code to be Coverity-proof as much as possible.

Reply:
No.... the control flow in the initial implementation of posix_spawn was flawed due to the extra `ret == -1` checks after every operation, and that should `return -1` in case it satisfies. But, here the control flow is fine and I don't think it will cause any issues in coverity scanner.